src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      stateCopy = addProperties(stateCopy, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      stateCopy = removeProperties(stateCopy, action.keysToRemove);
+    } else if (action.type === 'clear') {
+      stateCopy = clear(stateCopy);
+    }
+
+    stateCopies.push({ ...stateCopy });
+  }
+
+  return stateCopies;
+}
+
+function addProperties(stateCopy, extraData) {
+  return {
+    ...stateCopy,
+    ...extraData,
+  };
```

Comment:
The function name 'addProperties' does not clearly convey that it creates and returns a new object. Consider renaming it to something like 'createStateWithAddedProperties' to better reflect its functionality.